libraries/core_libs/consensus/src/vote_manager/vote_manager.cpp

```diff
@@ -537,26 +542,55 @@ std::optional<PbftRound> VoteManager::determineNewRound(PbftPeriod current_pbft_
 }
 
 PbftPeriod VoteManager::getRewardVotesPbftBlockPeriod() {
-  std::shared_lock lock(reward_votes_info_mutex_);
+  std::shared_lock lock(verified_votes_access_);
   return reward_votes_period_;
 }
 
-void VoteManager::resetRewardVotesInfo(PbftPeriod period, PbftRound round, const blk_hash_t& block_hash) {
-  {
-    std::scoped_lock lock(reward_votes_info_mutex_);
-
-    reward_votes_block_hash_ = block_hash;
-    reward_votes_period_ = period;
-    reward_votes_round_ = round;
+void VoteManager::resetRewardVotes(PbftPeriod period, PbftRound round, PbftStep step, const blk_hash_t& block_hash,
+                                   DbStorage::Batch& batch) {
+  // Save 2t+1 cert votes to database, remove old reward votes
+  std::scoped_lock lock(verified_votes_access_);
+  reward_votes_block_hash_ = block_hash;
+  reward_votes_period_ = period;
+  reward_votes_round_ = round;
+
+  auto found_period_it = verified_votes_.find(period);
+  // Add period
+  if (found_period_it != verified_votes_.end()) {
+    auto found_round_it = found_period_it->second.find(round);
+    // Add round
+    if (found_round_it != found_period_it->second.end()) {
+      auto found_step_it = found_round_it->second.step_votes.find(step);
+      // Add step
+      if (found_step_it != found_round_it->second.step_votes.end()) {
+        auto found_two_t_plus_one_voted_block =
+            found_round_it->second.two_t_plus_one_voted_blocks_.find(TwoTPlusOneVotedBlockType::CertVotedBlock);
+
+        // 2t+1 votes block already set
+        if (found_two_t_plus_one_voted_block != found_round_it->second.two_t_plus_one_voted_blocks_.end()) {
+          assert(found_two_t_plus_one_voted_block->second.first == block_hash);
+          auto found_voted_value_it = found_step_it->second.votes.find(block_hash);
+          // Add voted value
+          if (found_voted_value_it != found_step_it->second.votes.end()) {
+            std::vector<std::shared_ptr<Vote>> votes;
+            votes.reserve(found_voted_value_it->second.second.size());
+            for (const auto& tmp_vote : found_voted_value_it->second.second) {
```

Comment:
why do we need this if you can just send vector of cert votes from the place where we call resetRewardVotes ?

Reply:
There is a race condition. Some new votes might have arrived in the meantime and be stored in verified_votes and then we would lose those.